### PR TITLE
feat: add new health check runner package

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

We want the ability to monitor the output of simple "health" checks run periodically against smart contracts. This PR adds a very simple node script that can run sub-processes and log if any of them produce errors. In theory this can be used to monitor the correctness of execution of processes running in serverless and use this to run contract health and integrity checks.

### CODE REVIEW

File: packages/bot-strategy-runner/CHANGELOG.md [hunk 1]
```diff
@@ -1,242 +0,0 @@
-# Change Log
```
Comment (mrice32) [inline packages/bot-strategy-runner/CHANGELOG.md:1]: nit: what are all of these deletions? Are these related?

Comment (chrismaree) [reply packages/bot-strategy-runner/CHANGELOG.md:1]: > nit: what are all of these deletions? Are these related?

review this first: https://github.com/UMAprotocol/protocol/pull/4309

that's why this PR is in draft

File: packages/health-check-runner/index.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+const exec = require("child_process").exec;
```
Comment (chrismaree) [inline packages/health-check-runner/index.ts:1]: ^

File: packages/health-check-runner/index.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  });
+}
+
+// Format stderr outputs.
+function _stripExecStdOut(output) {
+  if (!output) return output;
+  /* eslint-disable no-control-regex */
+  return output
+    .replace(/\r?\n|\r|/g, "") // Remove Line Breaks, Reversing Strings
+    .replace(/[\u001b\u009b][[()#;?]*(?:[0-9]{1,4}(?:;[0-9]{0,4})*)?[0-9A-ORZcf-nqry=><]/g, ""); // Remove all ANSI colors/style
```
Comment (mrice32) [inline packages/health-check-runner/index.ts:60]: Why do we need this formatting?

For instance, for the colors and style, are there options or env vars that we can provide to the entire process to disable those?

File: packages/health-check-runner/index.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  const outputs = await Promise.all(healthCheckCommands.map(async (command) => execShellCommand(command, process.env)));
+  const errorOutputs = outputs.filter((output) => output.error);
+  const validOutputs = outputs.filter((output) => !output.error);
+  const outputLogLevel = errorOutputs.length > 0 ? "error" : "debug";
+
+  logger[outputLogLevel]({
+    at: "HealthCheckRunner",
+    message: `Health check command ${outputLogLevel == "error" ? "failed" : "succeeded"}!`,
+    validOutputs: validOutputs.map((output) => output.cmd),
+    errorOutputs,
```
Comment (mrice32) [inline packages/health-check-runner/index.ts:29]: Have we tested to ensure this list of errors stringifies correctly with the logger? Should we pre-stringify to avoid the risk of `[Object object]` type outputs?

File: packages/core/test/foundry/optimistic-asserter/OptimisticAsserter.Maintenance.t.sol [hunk 5]
```diff
@@ -100,5 +100,6 @@ contract MaintenanceTest is Common {
         );
         vm.stopPrank();
         assertTrue(optimisticAsserter.cachedIdentifiers(newIdentifier));
+        assertTrue(false);
```
Comment (mrice32) [inline packages/core/test/foundry/optimistic-asserter/OptimisticAsserter.Maintenance.t.sol:103]: nit: what is this change?

File: packages/health-check-runner/index.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
+const exec = require("child_process").exec;
+import { Logger } from "@uma/financial-templates-lib";
+
+let logger;
+
+export async function run(): Promise<void> {
```
Comment (mrice32) [inline packages/health-check-runner/index.ts:6]: Just OOC, why couldn't we run all of the inner health check commands directly from the Hub/Spoke rather than batching them in here? As far as I can see, they're just generic node commands, and the Spoke already knows how to call those, right?

**Review by mrice32 — COMMENTED** [review]
Looks great! Just a few comments.

File: packages/core/test/foundry/optimistic-asserter/OptimisticAsserter.Maintenance.t.sol (hunk 5 above)
Comment (chrismaree) [reply packages/core/test/foundry/optimistic-asserter/OptimisticAsserter.Maintenance.t.sol]: this was testing that things produced the correct error. the PR is still in draft ;)

File: packages/health-check-runner/index.ts (hunk 6 above)
Comment (chrismaree) [reply packages/health-check-runner/index.ts:6]: this is already going to run in the hub and spoke but it lets us define _one_ serverless job to run a set of "health checks", rather than needing to have each check be it's own spoke job. the main benefits are in the logging: we don't need to worry about how the health checks log: they just need to produce: a) exit codes relating to their output and b) console logs (standard out). this runner then executes them and wraps the logic to group logs together and process the exit codes.

File: packages/health-check-runner/index.ts (hunk 3 above)
Comment (chrismaree) [reply packages/health-check-runner/index.ts:60]: foundry appends color to the output of it's logs that makes them looks really nasty if you extract the standard out from the process. doing this means that they they don't look too bad, when read in slack or in GCP.

File: packages/health-check-runner/index.ts (hunk 4 above)
Comment (chrismaree) [reply packages/health-check-runner/index.ts:29]: I've tested it running hardhat scripts that error if there is a problem and foundry tests against a mainnet fork.

**Review by md0x — APPROVED** [review]
LGTM!

File: packages/core/package.json [hunk 7]
```diff
@@ -8,6 +8,7 @@
     "@maticnetwork/fx-portal": "^1.0.4",
     "@openzeppelin/contracts": "4.4.2",
     "@uma/common": "^2.28.4",
+    "@uma/merkle-distributor": "^1.3.38",
```
Comment (chrismaree) [inline packages/core/package.json:11]: This is needed to address a CI issue present from the most recent removal of packages from the repo.

File: packages/health-check-runner/index.ts (hunk 3 above)
Comment (mrice32) [reply packages/health-check-runner/index.ts:60]: There's no way to disable them?

File: packages/core/test/foundry/optimistic-asserter/OptimisticAsserter.Maintenance.t.sol (hunk 5 above)
Comment (mrice32) [reply packages/core/test/foundry/optimistic-asserter/OptimisticAsserter.Maintenance.t.sol]: No, no worries, just didn't know if it was intentional or not! Makes sense. Disregard :)

File: packages/health-check-runner/index.ts (hunk 6 above)
Comment (mrice32) [reply packages/health-check-runner/index.ts:6]: I totally hear you. Only pushing on reducing layers/complexity, since this code feels pretty similar to our spoke code. I think one of the issues we've had in the past is that fanciness around log processing and layers of subprocesses has resulted in a) fewer people understanding the architecture/what's happening and b) more bugs/maintainability issues in the code.

I don't have a strong opinion here, and I can see the argument for doing it both ways, I just want to make sure we're thinking about the layering complexity and considering alternatives!

File: packages/health-check-runner/index.ts (hunk 4 above)
Comment (mrice32) [reply packages/health-check-runner/index.ts:29]: Sounds good, as long as you're confident the error list is _always_ stringified in a readable way, this SGTM!

This is probably the most important output of this, so it's super important that the person being notified can read the errors and all of their properties no matter the environment where it's run.

**Review by mrice32 — APPROVED** [review]
Responded to the comments with some additional thoughts!

I would really like to avoid complicated log processing and regexes since those have bitten us in the past, but I totally get that sometimes those are unavoidable in our serverless env.
